genie-web/src/main/java/com/netflix/genie/web/agent/services/impl/AgentConnectionTrackingServiceImpl.java

```diff
@@ -0,0 +1,214 @@
+package com.netflix.genie.web.agent.services.impl;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+import com.netflix.genie.web.agent.services.AgentConnectionTrackingService;
+import com.netflix.genie.web.agent.services.AgentRoutingService;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.scheduling.TaskScheduler;
+
+import java.time.Instant;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Supplier;
+
+/**
+ * This service keeps track of agent connections and heartbeats. It notifies the downstream {@link AgentRoutingService}
+ * of connected/disconnected agents while hiding details of connections, disconnections, missed heartbeats.
+ *
+ * @author mprimi
+ * @since 4.0.0
+ */
+@Slf4j
+public class AgentConnectionTrackingServiceImpl implements AgentConnectionTrackingService {
+
+    // Minimum time from the last heartbeat before a stream is considered expired
+    protected static final int STREAM_EXPIRATION_PERIOD = 10_000; // TODO: Make configurable
+    protected static final int CLEANUP_TASK_PERIOD = 2_000; // TODO: Make configurable
+    private final AgentRoutingService agentRoutingService;
+    private final TaskScheduler taskScheduler;
+    private final ConcurrentMap<String, JobStreamsRecord> jobStreamRecordsMap = Maps.newConcurrentMap();
+    private final Supplier<Instant> timeSupplier;
+
+    /**
+     * Constructor.
+     *
+     * @param agentRoutingService the agent routing service
+     * @param taskScheduler       the task scheduler
+     */
+    public AgentConnectionTrackingServiceImpl(
+        final AgentRoutingService agentRoutingService,
+        final TaskScheduler taskScheduler
+    ) {
+        this(agentRoutingService, taskScheduler, Instant::now);
+    }
+
+    @VisibleForTesting
+    AgentConnectionTrackingServiceImpl(
+        final AgentRoutingService agentRoutingService,
+        final TaskScheduler taskScheduler,
+        final Supplier<Instant> timeSupplier
+    ) {
+        this.agentRoutingService = agentRoutingService;
+        this.taskScheduler = taskScheduler;
+        this.timeSupplier = timeSupplier;
+
+        this.taskScheduler.scheduleAtFixedRate(this::cleanupTask, CLEANUP_TASK_PERIOD);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void notifyHeartbeat(final String streamId, final String claimedJobId) {
+        final AtomicBoolean isNew = new AtomicBoolean(false);
+
+        final JobStreamsRecord jobStreamsRecord = this.jobStreamRecordsMap.computeIfAbsent(
+            claimedJobId,
+            jobId -> {
+                isNew.set(true);
+                return new JobStreamsRecord(jobId);
+            }
+        );
+
+        log.debug(
+            "Received heartbeat for {} job {} using stream {}",
+            isNew.get() ? "new" : "existing",
```

Comment:
Couple of minor things. Could make strings constants and `isNew.get()` is called multiple. times in this method. You could just unwrap it once after the compute if absent call